panel/widgets/terminal.py

```diff
@@ -163,25 +163,28 @@ def _reset(self):
     def _remove_last_line_from_string(value):
         return value[: value.rfind("CompletedProcess")]
 
+    def _decode_utf8_on_boundary(self, fd, max_read_bytes):
+        "UTF-8 characters can be multi-byte so need to decode on correct boundary"
+        raw = os.read(fd, max_read_bytes)
+        try:
+            return raw.decode('utf-8')
+        except UnicodeDecodeError:
+            extra_byte = os.read(fd, 1)
+            data = (raw + extra_byte)
+            try:
+                return data.decode('utf-8')
+            except UnicodeDecodeError:
+                another_extra_byte = os.read(fd, 1)
+                return (data + another_extra_byte).decode('utf-8')
```

Comment:
```suggestion
    def _decode_utf8_on_boundary(self, fd, max_read_bytes, max_retries=2):
        "UTF-8 characters can be multi-byte so need to decode on correct boundary"
        raw = os.read(fd, max_read_bytes)
        while max_retries > 0:
            try:
                decoded = raw.decode('utf-8')
            except UnicodeDecodeError:
                raw = (raw + os.read(fd, 1))
                max_retries -= 1
                continue
            return decoded
        raise UnicodeDecodeError('Could not utf-8 decode subprocess output.')
```

Reply:
I've implemented this differently already (and the `UnicodeDecodeError` at the end has a different signature, it doesn't behave like a normal exception)